### PR TITLE
Enhance `no_grad`-context FSDP backward handling

### LINKED ISSUE #105369: FSDP post-backward reshard-only hooks registered in `no_grad` context
### 🐛 Describe the bug

FSDP's `_post_backward_reshard` (reshard-only) hooks are currently registered in `no_grad` contexts. As the hook cleanup code-paths are not executed in `no_grad` contexts, the first `grad` mode step succeeding a `no_grad` step can encounter the below `SystemError` on `run_backward`:

```python
./test_fsdp_fine_tune.py::TestFSDPFineTune::test_backward_reshard_hooks Failed with Error: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_distributed.py", line 657, in run_test
    getattr(self, test_name)()
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_distributed.py", line 543, in wrapper
    fn()
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_utils.py", line 2354, in wrapper
    method(*args, **kwargs)
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_distributed.py", line 174, in wrapper
    return func(*args, **kwargs)
  File "/home/speediedan/repos/pytorch/test/distributed/fsdp/test_fsdp_fine_tune.py", line 65, in test_backward_reshard_hooks
    self.run_subtests(
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_fsdp.py", line 896, in run_subtests
    return run_subtests(self, *args, **kwargs)
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_fsdp.py", line 848, in run_subtests
    test_fn(*test_args, **test_kwargs, **subtest_kwargs)
  File "/home/speediedan/repos/pytorch/test/distributed/fsdp/test_fsdp_fine_tune.py", line 142, in _test_backward_reshard_hooks
    output.sum().backward()
  File "/home/speediedan/repos/pytorch/torch/_tensor.py", line 491, in backward
    torch.autograd.backward(
  File "/home/speediedan/repos/pytorch/torch/autograd/__init__.py", line 204, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
SystemError: <built-in method run_backward of torch._C._EngineBase object at 0x7fe3fe8df8a0> returned NULL without setting an exception
```

### To reproduce

In the PR I will be submitting shortly, I am extending the existing [`test_fsdp_fine_tune.py::TestFSDPFineTune::test_backward_reshard_hooks`](https://github.com/pytorch/pytorch/blob/93f852f201b93ca0c41b5cd861834d4f1f235ef7/test/distributed/fsdp/test_fsdp_fine_tune.py#L79) test to interleave a `no_grad` step along with subsequent post-backward hook count (and `flat_param` `requires_grad`) validation.

### Versions

PyTorch version: 2.1.0.dev20230717
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.31

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-152-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 4090
GPU 1: NVIDIA GeForce RTX 2070 SUPER

Nvidia driver version: 530.41.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.2
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          24
On-line CPU(s) list:             0-23
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           1
Model name:                      AMD Ryzen Threadripper 1920X 12-Core Processor
Stepping:                        1
Frequency boost:                 enabled
CPU MHz:                         2107.217
CPU max MHz:                     3500.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        6985.93
Virtualization:                  AMD-V
L1d cache:                       384 KiB
L1i cache:                       768 KiB
L2 cache:                        6 MiB
L3 cache:                        32 MiB
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate sme ssbd sev ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h6d00ec8_46342  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.6           py310h1128e8f_1  
[conda] mkl_random                1.2.2           py310h1128e8f_1  
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchtriton               2.1.0+3c400e7818           py310    pytorch-nightly

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### LINKED ISSUE #105371: `NO_RESHARD_AFTER_FORWARD_HANDLE_STRATEGIES` strategies may fail to reset `flat_param` `requires_grad` after `requires_grad` changed on original parameters
### 🐛 Describe the bug

With the existing [`test_fsdp_fine_tune.py::TestFSDPFineTune::test_backward_reshard_hooks`](https://github.com/pytorch/pytorch/blob/93f852f201b93ca0c41b5cd861834d4f1f235ef7/test/distributed/fsdp/test_fsdp_fine_tune.py#L79) test, adding a `no_grad` step results in subsequent `requires_grad` changes being lost. 

Because the `SHARD_GRAD_OP` strategy does not re-shard parameters post-forward, changes to `requires_grad` on original parameters will be lost in the `grad` step succeeding a `no_grad` step (since `_use_sharded views` is skipped). Of course, this is a relatively common pattern, e.g.:
1. train for `n` steps
2. validation step executed (often at end of each epoch)
3. unfreeze additional parameters and continue training

In the PR I will be submitting shortly, I extend [`test_fsdp_fine_tune.py::TestFSDPFineTune::test_backward_reshard_hooks`](https://github.com/pytorch/pytorch/blob/93f852f201b93ca0c41b5cd861834d4f1f235ef7/test/distributed/fsdp/test_fsdp_fine_tune.py#L79) to validate appropriate `flat_param` `requires_grad` changes in `grad` steps succeeding `no_grad` steps.

One potential remediation (in my forthcoming PR) is to, only in a `no_grad` context, switch back to using sharded views in `_use_sharded_flat_param` (called in `_reshard` codepath) to allow subsequent step's `_reset_flat_param_grad_info_if_needed` invocations to update flat param `requires_grad` using the appropriate original parameter views.

### To reproduce

In the PR I will be submitting shortly, I am extending the existing [`test_fsdp_fine_tune.py::TestFSDPFineTune::test_backward_reshard_hooks`](https://github.com/pytorch/pytorch/blob/93f852f201b93ca0c41b5cd861834d4f1f235ef7/test/distributed/fsdp/test_fsdp_fine_tune.py#L79) test to interleave a `no_grad` step along with subsequent `flat_param` `requires_grad` (and post-backward hook count) validation.

```python
./test_fsdp_fine_tune.py::TestFSDPFineTune::test_backward_reshard_hooks Failed with Error: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_distributed.py", line 657, in run_test
    getattr(self, test_name)()
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_distributed.py", line 543, in wrapper
    fn()
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_utils.py", line 2354, in wrapper
    method(*args, **kwargs)
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_distributed.py", line 174, in wrapper
    return func(*args, **kwargs)
  File "/home/speediedan/repos/pytorch/test/distributed/fsdp/test_fsdp_fine_tune.py", line 65, in test_backward_reshard_hooks
    self.run_subtests(
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_fsdp.py", line 896, in run_subtests
    return run_subtests(self, *args, **kwargs)
  File "/home/speediedan/repos/pytorch/torch/testing/_internal/common_fsdp.py", line 848, in run_subtests
    test_fn(*test_args, **test_kwargs, **subtest_kwargs)
  File "/home/speediedan/repos/pytorch/test/distributed/fsdp/test_fsdp_fine_tune.py", line 148, in _test_backward_reshard_hooks
    _assert_post_backward_requires_grad(seq)
  File "/home/speediedan/repos/pytorch/test/distributed/fsdp/test_fsdp_fine_tune.py", line 106, in _assert_post_backward_requires_grad
    self.assertTrue(all(p.requires_grad for p in seq.parameters()),
  File "/opt/miniconda/envs/pytorch_nightly/lib/python3.10/unittest/case.py", line 687, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true : Expected all parameters to require grad but some did not!
```

### Versions

PyTorch version: 2.1.0.dev20230717
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.31

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-152-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 4090
GPU 1: NVIDIA GeForce RTX 2070 SUPER

Nvidia driver version: 530.41.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.2
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          24
On-line CPU(s) list:             0-23
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           1
Model name:                      AMD Ryzen Threadripper 1920X 12-Core Processor
Stepping:                        1
Frequency boost:                 enabled
CPU MHz:                         2107.217
CPU max MHz:                     3500.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        6985.93
Virtualization:                  AMD-V
L1d cache:                       384 KiB
L1i cache:                       768 KiB
L2 cache:                        6 MiB
L3 cache:                        32 MiB
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate sme ssbd sev ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h6d00ec8_46342  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.6           py310h1128e8f_1  
[conda] mkl_random                1.2.2           py310h1128e8f_1  
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchtriton               2.1.0+3c400e7818           py310    pytorch-nightly

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### PR DESCRIPTION
**Pull request by speediedan**
Fixes #105369
Fixes #105371

Addressing two somewhat distinct issues that involve the same test in this PR:

1. To fix #105369:
    - Add a `no_grad` guard to [`_register_post_backward_reshard_only_hooks`](https://github.com/pytorch/pytorch/blob/93f852f201b93ca0c41b5cd861834d4f1f235ef7/torch/distributed/fsdp/_runtime_utils.py#L1406) to avoid registering post-backward hooks that would not be removed in that context.

2. To fix #105371:
    - Add a `grad` context condition to [`_use_sharded_flat_param`](https://github.com/pytorch/pytorch/blob/93f852f201b93ca0c41b5cd861834d4f1f235ef7/torch/distributed/fsdp/flat_param.py#L1645C9-L1645C32) logic to trigger post-forward `_use_sharded_views` in a `no_grad` context for `NO_RESHARD_AFTER_FORWARD_HANDLE_STRATEGIES`

### CODE REVIEW

**Review by awgu — COMMENTED** [review]
I think the change to `_runtime_utils.py` is definitely reasonable.

For the change in `flat_param.py`, it feels like a patch, but maybe it is reasonable. I feel a bit regretful about adding the logic for skipping `_use_sharded_views()` for `SHARD_GRAD_OP` in pursuit of reducing CPU overhead since it introduces so much special handling. Overall, I am okay with this change in that it simply makes the criteria more strict for when we can save on that CPU overhead.

If skipping `_use_sharded_views()` continues to be a problem, then we may need to revert it, or do a deeper refactor.

**Comment by speediedan** [comment]
> I think the change to `_runtime_utils.py` is definitely reasonable.
> 
> For the change in `flat_param.py`, it feels like a patch, but maybe it is reasonable. I feel a bit regretful about adding the logic for skipping `_use_sharded_views()` for `SHARD_GRAD_OP` in pursuit of reducing CPU overhead since it introduces so much special handling. Overall, I am okay with this change in that it simply makes the criteria more strict for when we can save on that CPU overhead.
> 
> If skipping `_use_sharded_views()` continues to be a problem, then we may need to revert it, or do a deeper refactor.

Totally agree regarding that simplicity vs resource efficiency tension. Thanks for the timely and thoughtful review!

**Comment by speediedan** [comment]
Resolved conflict as `flat_param.py` change no longer necessary after @awgu's nice refactor in #105346. Hope this will be ready to merge now. Thanks @awgu!

**Review by awgu — APPROVED** [review]
This looks good to me. Thanks for this!

File: test/distributed/fsdp/test_fsdp_fine_tune.py [hunk 1]
```diff
@@ -101,11 +101,36 @@ def _post_backward_reshard_with_count(*args, **kwargs):
             post_backward_reshard_count += 1
             return orig_post_backward_reshard(*args, **kwargs)
 
+        def _assert_post_backward_requires_grad(seq):
+            if step_idx == num_steps - 1 and unfreeze_params:
+                self.assertTrue(
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_fine_tune.py:106]: It looks like this assert is failing.

Comment (speediedan) [reply test/distributed/fsdp/test_fsdp_fine_tune.py:106]: Sure thing! When I merged the conflict with #105346 I didn't have a chance to pull that PR down and test locally first. In the future, even if I expect tests to pass, I'll be more prudent and save CI resources by pulling conflicting PRs down and testing locally before resolving conflicts. 

Just had a chance to take a quick look locally and validated that if we return to using the somewhat more constrained criteria [we discussed](https://github.com/pytorch/pytorch/pull/105374#pullrequestreview-1533908546) for skipping use of sharded views, the test passes again, i.e. altering
https://github.com/pytorch/pytorch/blob/8bd8615ff14e66ce3e1bcc0647b4408baa5aa339/torch/distributed/fsdp/flat_param.py#L1657-L1666
with the discussed additional grad condition:
```python
            skip_use_sharded_views = (
                torch.is_grad_enabled()
                and in_forward
                and self._sharding_strategy
                in NO_RESHARD_AFTER_FORWARD_HANDLE_STRATEGIES
            )
````
Let me know if you're still okay continuing with that approach and I'll push that commit. Thanks!

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_fine_tune.py:106]: Yes, that approach sounds good to me.

**Comment by speediedan** [comment]
@awgu any sense of when this can be merged? I know the team has a lot on its plate with some exciting changes on the horizon, so just hoping to avoid future merge conflicts and ensure this PR doesn't drop off the radar. Thanks!

**Review by awgu — APPROVED** [review]
Sorry, I missed that you updated this. Thanks for bumping this.

Thanks for the fix!

**Comment by speediedan** [comment]
> Thanks for bumping this.
> 
> Thanks for the fix!

A pleasure as always!

**Comment by speediedan** [comment]
[current failure](https://github.com/pytorch/pytorch/actions/runs/5660499951/job/15336917773?pr=105374#step:11:115337) appears unrelated so we should be able to force merge:
```
2023-07-25T23:11:07.8261578Z FAILED [0.0843s] test_sort_and_select.py::TestSortAndSelectCUDA::test_sort_large_cuda_uint8 - torch.cuda.OutOfMemoryError: HIP out of memory. Tried to allocate 16.00 GiB. GPU 0 has a total capacty of 15.98 GiB of which 131059.84 GiB is free...
```

**Pull request closed without merging**